### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by maasencioh**
Find the nearest point to a sample point

cc @mljs/collaborators

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,5 +1,61 @@
...
+};
+
+/**
+ *
+ * @param {Array<Array<Number>>} listVectors - List of vectors with same dimensions
+ * @param {Array<Number>} vector - Reference vector to "classify"
+ * @param {Object} [options] - Options object
+ * @param {Function} [options.distanceFunction = squaredDistance] - Function that receives two vectors and return their distance value as number
+ * @param {Function} [options.similarityFunction = undefined] - Function that receives two vectors and return their similarity value as number
+ * @param {Boolean} [options.returnObject = false] - Return the nearest vector instead of the index of it
```
Comment (targos) [inline src/index.js]: "instead of its index"

File: src/index.js [hunk 2]
```diff
@@ -1,5 +1,61 @@
...
+ * @return {Number|Array<Number>} - The index or the content of the nearest vector
+ */
+function nearestVector(listVectors, vector, options) {
+    options = options || defaultOptions;
+    options.distanceFunction = options.distanceFunction || defaultOptions.distanceFunction;
```
Comment (targos) [inline src/index.js]: I wouldn't override the values in the options. Just create local variables:

```
var distanceFunction = options.distanceFunction || defaultOptions.distanceFunction;
// ...
```

File: src/index.js [hunk 3]
```diff
@@ -1,5 +1,61 @@
...
+    options.similarityFunction = options.similarityFunction || defaultOptions.similarityFunction;
+    options.returnObject = options.returnObject || defaultOptions.returnObject;
+
+    var vectorIndex = -1;
+    if (typeof options.similarityFunction === 'function') {
+
+        // maximum similarity
+        var maxSim = -1;
```
Comment (targos) [inline src/index.js]: Number.MIN_VALUE ?

File: src/index.js [hunk 4]
```diff
@@ -1,5 +1,61 @@
...
 
-function nearestVector() {}
+const squaredDistance = require('ml-distance-euclidean').squared;
 
-module.exports = nearestVector();
\ No newline at end of file
+const defaultOptions = {
+    distanceFunction: squaredDistance,
+    similarityFunction: false,
+    returnObject: false
```
Comment (targos) [inline src/index.js]: returnVector ?

File: src/index.js (hunk 3 above)
Comment (maasencioh) [reply src/index.js]: A similarity should be defined in the set [0, 1], but It's not a problem to change that

**Comment by andcastillo** [comment]
I have 2 observations for this project:
1. I would extended this function to get the k-nearest-neighbours. 
2. In second place, I would allow a parameter/option that is an order-function and I would not differentiate it from a distance or similarity function. It should be a function that induces and order, and you will get the k-first elements of the resulting vector.

**Comment by targos** [comment]
that would be for another project. we need this one to be simple and fast for use in the kmeans algorithm and image-js

**Comment by maasencioh** [comment]
@andcastillo I added the first one as an issue, I like the idea, but for the second what could be the difference to the [Array.sort](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) property?

**Comment by targos** [comment]
remember: in Node.js modules are recommended to be as single-purpose as possible for reusability

**Comment by andcastillo** [comment]
@maasencioh the difference is that you can get k-nearest-neighbours without explicitly make a sort of your whole array. You only remember the k-elements that you need, which will save a lot of memory.
@targos, if instead of 2 different kind of function you use a single one, the code gains in simplicity and then it become a "single-purpose" function. It is the user who takes care of the function that gives to produce the result, and then we can re-use easily, because I think we will soon need not only the k-means, but also k-nearest-neighbour algorithm.

**Comment by maasencioh** [comment]
@andcastillo we already have [k-nearest-neighbour](https://github.com/mljs/knn) and now I get what you want, but for that will be more interesting to create a sort of k-quickselect algortihm, based on [quickselect](https://en.wikipedia.org/wiki/Quickselect)

**Comment by andcastillo** [comment]
Now, I'm understand. The name nearest-vector sound to me to nearest-neighbour. It is not a vector, it is a tuple, if you have non-numeric values.  It's ok, if you need only the 1-nearest "vector" the code is ok.

**Pull request merged**